crates/roc_std/src/roc_str.rs

```diff
@@ -182,34 +211,91 @@ impl RocStr {
 
             if target_cap > SmallString::CAPACITY {
                 // The requested capacity won't fit in a small string; we need to go big.
-                let mut roc_list = RocList::with_capacity(target_cap);
+                let mut big_string = BigString::with_capacity(target_cap);
 
-                roc_list.extend_from_slice(small_str.as_bytes());
-
-                *self = RocStr(RocStrInner {
-                    heap_allocated: ManuallyDrop::new(roc_list),
+                unsafe {
+                    std::ptr::copy_nonoverlapping(
+                        self.as_bytes().as_ptr(),
+                        big_string.ptr_to_first_elem(),
+                        self.len(),
+                    )
+                };
+
+                big_string.length = self.len();
+                big_string.capacity_or_ref_ptr = target_cap;
+
+                let mut updated = RocStr(RocStrInner {
+                    heap_allocated: ManuallyDrop::new(big_string),
                 });
+
+                mem::swap(self, &mut updated);
+                mem::forget(updated);
             }
         } else {
-            let mut roc_list = unsafe { ManuallyDrop::take(&mut self.0.heap_allocated) };
+            let mut big_string = unsafe { ManuallyDrop::take(&mut self.0.heap_allocated) };
 
-            roc_list.reserve(bytes);
+            big_string.reserve(bytes);
 
             let mut updated = RocStr(RocStrInner {
-                heap_allocated: ManuallyDrop::new(roc_list),
+                heap_allocated: ManuallyDrop::new(big_string),
             });
 
             mem::swap(self, &mut updated);
             mem::forget(updated);
         }
     }
 
+    #[track_caller]
+    pub fn slice_range(&self, range: Range<usize>) -> Self {
+        match self.try_slice_range(range) {
+            Some(x) => x,
+            None => panic!("slice index out of range"),
+        }
+    }
+
+    pub fn try_slice_range(&self, range: Range<usize>) -> Option<Self> {
+        if self.as_str().get(range.start..range.end).is_none() {
+            None
+        } else if range.end - range.start <= SmallString::CAPACITY {
+            let slice = &self.as_bytes()[range];
+            let small_string =
+                unsafe { SmallString::try_from_utf8_bytes(slice).unwrap_unchecked() };
+
+            // NOTE decrements `self`
+            Some(RocStr(RocStrInner { small_string }))
+        } else {
+            // increment the refcount
+            std::mem::forget(self.clone());
+
+            let big = unsafe { &self.0.heap_allocated };
+            let ptr = unsafe { (self.as_bytes().as_ptr() as *mut u8).add(range.start) };
+
+            let heap_allocated = ManuallyDrop::new(BigString {
+                elements: unsafe { NonNull::new_unchecked(ptr) },
+                length: (isize::MIN as usize) | (range.end - range.start),
+                capacity_or_ref_ptr: (big.ptr_to_first_elem() as usize) >> 1,
```

Comment:
Not important for this pr, but I really want to rename everything call capacity_or_ref_ptr. It isn't that. It is capacity or the pointer to the original list bytes. Just not sure the best name for it yet. `capacity_or_allocation_ptr`, `capacity_or_original_ptr`, `capacity_or_base_ptr`?

I just find a lot of code confusing cause we picked `capacity_or_ref_ptr`. Just like in zig `getRefcountPtr` actually returns the pointer after the refcount. Just bad naming that leads to confusing code.